WikiWatershed/model-my-watershed
# Hydroshare Export Login Backend

## Overview

In order to interface with HydroShare, to create projects and so on, users need to connect their MMW accounts to their HydroShare accounts. HydroShare supports OAuth 2 for access authorization, which we use to retrieve an `access_token`. This `access_token` must be provided in all requests to the HydroShare REST API in order to read and write privileged items.

In this PR, backend endpoints are created that support retrieving `access_token` from HydroShare for an MMW user and saving these to the database linked to said user. Model and serializers for saving the tokens to the database are also added. Environment variables for carrying client secrets needed for OAuth are established and added to the provisioning process. Finally, the [hs_restclient](https://github.com/hydroshare/hs_restclient) library is added for interacting with the HydroShare API.

Next steps will be to add backend endpoints for exporting projects to HydroShare, and adding frontends for connecting an MMW account to HydroShare using the endpoints created herein.

Connects #2513 

### Demo

![2017-11-22 11 31 37](https://user-images.githubusercontent.com/1430060/33138659-0e07ca84-cf79-11e7-8239-bb8ff6567762.gif)

### Notes

There are two commits with `TODO` comments:

 * ab14313 notes where it would be useful to have additional checking of input. Since that checking may be done a level above, I have left it unimplemented. But if we'd like more safety I can add it in.
 * e7cdc17 is a demo commit, and should be dropped from this PR before merging.

We also need to add the client secret for staging and production to the Terraform setup files used in deployment. Will need to pair with @tnation14 for that. I've requested his review for the deployment / provisioning parts of this PR as well.

## Testing Instructions

* Check out this branch. Run migrations.
* Find `HydroShare OAuth MMW Staging` credentials in LastPass, open the details view, and copy the `client-secret` value from it. Using it, reprovision `app`:

      (bash) $ MMW_HYDROSHARE_SECRET_KEY=*** vagrant reload app --provision
      (fish) > env MMW_HYDROSHARE_SECRET_KEY=*** vagrant reload app --provision

* Go to [:8000/](http://localhost:8000/). Make sure you are logged in.
* Go to [:8000/user/hydroshare/login](http://localhost:8000/user/hydroshare/login). You will be redirected to [dev.hydroshare.org](https://dev.hydroshare.org/).
* Log in as `model-my-watershed`. Find credentials under `HydroShare OAuth MMW Staging` in LastPass.
* Authorize `model-my-watershed` app to read and write to your account. You will be redirected to [:8000/user/hydroshare/authorize?code=XYZ](http://localhost:8000/user/hydroshare/authorize). You should see the access token details in this page.
* Go to [:5433/](http://localhost:5433/) to open pgweb and find the `user_hydrosharetoken` table and view its contents. Ensure the access token details have been recorded there.
* Go to [:8000/user/hydroshare/refresh](http://localhost:8000/user/hydroshare/refresh) and ensure you see a new access token and refresh token, without having to reauthorize (as you would if you go to [:8000/user/hydroshare/login](http://localhost:8000/user/hydroshare/login)). Ensure these new values are saved to the database.
* Go to [:8000/user/hydroshare/example](http://localhost:8000/user/hydroshare/example) and ensure you see exactly one "Test Resource" created under the `model-my-watershed` account. This is being pulled via the [hs_restclient](https://github.com/hydroshare/hs_restclient) from [dev.hydroshare.org](https://dev.hydroshare.org/) using the access token for `model-my-watershed`.
